userbindmount.1
.TH VDENS 1 "August 22, 2017" "VirtualSquare Labs"
.SH NAME
userbindmount \- bind-mount utility for user-namespaces
.SH SYNOPSIS
.B userbindmount
.RI "[options] [" source " " target " [" source " " target " [...]]] [ " -- " [" cmd " [" args "]]]"

.SH DESCRIPTION

userbindmount is a utility command based on libuserbindmount.

It can be used to perform one or more bind-mount operation and to create a user-namespace where
bind-mount is allowed.

This command does not need root access or specific capabilities to run (provided user-namespaces are supported,
see NOTES).

The command line arguments are a list of \fIsource\fR-\fItarget\fR pairs (one for each
bind-mount operation). 
A new namespace is created if requested by the specific option (-n or
--newns) or if there is -- as an option in the command line.
If \fIsource\fR is a double quoted string, the value of the string will be the content
of the file mounted on \fItarget\fR.
The trailing -- followed by a command and its argument define the command
to run in the new namespace. ($SHELL is launched if the command is omitted)

The contents of the file to be mounted on \fItarget\fR is read from
the standard input if the correspondent \fIsource\fR is the tag "-".

.SH OPTIONS
.I userbindmount
accepts the following options.

.TP
\fB\-n 
.TQ
\fB\-\-newns
create a new user-namespace

.TP
\fB\-s 
.TQ
\fB\-\-sysadm 
add the CAP_SYS_ADMIN ambient capability to the current of newly created userspace

.TP
\fB\-v 
.TQ
\fB\-\-verbose 
verbose mode: print debugging information on the actions taken by the program.

.SH NOTES
User namespaces require a kernel that is configured with the CONFIG_USER_NS option.
In some distributions (e.g. Debian) user namespaces must be enabled by writing 1 to
/proc/sys/kernel/unprivileged_userns_clone.
.SH EXAMPLES
The following example mounts the file /tmp/resolv.conf instead of /etc/resolv.conf: the purpose of 
this example is to redefine the name servers for the name resolution.
.RS
.nf
$ cat /etc/resolv.conf
nameserver 127.0.0.1
$ echo "nameserver 9.9.9.9" > /tmp/resolv.conf
$ userbindmount -v /tmp/resolv.conf /etc/resolv.conf -- bash
creating a user_namespace
mounting /tmp/resolv.conf on /etc/resolv.conf
starting bash
$ cat /etc/resolv.conf 
nameserver 9.9.9.9
$ exit
$
.fi
.RE

.sp
The following example creates a namespace where bind-mount is allowed and then mounts /tmp/resolv.conf on
/etc/resolv.conf. (It uses busybox instead of mount(8) as the latter does not support the capabilities, yet).
.RS
.nf 
$ userbindmount -s -- bash
$ cat /etc/resolv.conf 
nameserver 127.0.0.1
$ echo "nameserver 9.9.9.9" > /tmp/resolv.conf
$ busybox mount --bind /tmp/resolv.conf /etc/resolv.conf 
$ cat /etc/resolv.conf
nameserver 9.9.9.9
$ exit
$
.fi
.RE
Alternative equivalent commands for "userbindmount -s -- bash" are "userbindmount -sn" or "userbindmount -s --\fR".

.sp
Several bind-mounts can be done in a user-namespace started with the -s option. 
No more namespaces are needed in this case.
The contents of the file to mount can be taken from stdin if source is "-".
.RS
.nf
$ userbindmount -sn
$ echo "nameserver 9.9.9.9" | userbindmount - /etc/resolv.conf
$ cat /etc/resolv.conf
nameserver 9.9.9.9
$ exit
.fi
.RE

.sp
It is possible to set the contents of a mounted file directly in the command line:
.RS
.nf
$ userbindmount $'"nameserver 9.9.9.9\n"' /etc/resolv.conf -- bash
$ cat /etc/resolv.conf
nameserver 9.9.9.9
$ exit
.fi
.RE


Please note that the following command:
.RS
.nf
$ echo "nameserver 9.9.9.9" | userbindmount - /etc/resolv.conf -- bash
.fi
.RE
works but the bash running in the new namespace terminates immediately as it reads the end-of-file on its
standard input.
.SH SEE ALSO
.BR "libuserbindmount"(3), " mount"(8), " user_namespaces"(7), " capabilities"(7)
.SH BUGS
Bug reports should be addressed to <info@virtualsquare.org>
.SH AUTHORS
Renzo Davoli <renzo@cs.unibo.it>


userbindmount.c
/* 
 * userbindmount: bindmount as a user (using user-namespaces).
 * Copyright (C) 2016  Renzo Davoli, University of Bologna
 * 
 * userbindmount is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; If not, see <http://www.gnu.org/licenses/>. 
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <getopt.h>
#include <errno.h>
#include <libgen.h>
#include <sys/mount.h>

#include "userbindmount.h"

char *progname;
int sysadmin_flag = 0;
int verbose_flag = 0;
int newnamespace_flag = 0;

#define errExit(msg)    ({ perror(msg); exit(EXIT_FAILURE); })

void usage_n_exit(void) {
	fprintf(stderr, 
			"Usage: %s OPTIONS [source target [source target [...]]] [-- [cmd [args]]] \n"
			"OPTIONS:\n"
			"\t-n | --newns   : create a new namespace\n"
			"\t-s | --sysadm  : set cap_sys_admin ambient capability\n"
			"\t-v | --verbose : verbose mode\n" 
			"\t-h | --help    : print this short help message\n" 
			"a new user_namespace is created if -- arg is present\n"
			"it runs $SHELL if cmd omitted\n"
			"the contents of the mounted file is read from stdin if source is '-'\n\n",
			progname);
	exit(EXIT_SUCCESS);
}

void userbindmount_mount(const char *source, const char *target) {
	int retval;
	if (*source == 0 || strcmp(source, "-") == 0)
		retval = userbindmount_fd(0, target, 0600);
	else {
		int len = strlen(source);
		if (source[0] == '"' && len > 1 && source[len - 1] == '"') {
			char string[len];
			snprintf(string, len, "%*.*s", len - 2, len - 2, source + 1);
			retval = userbindmount_string(string, target, 0600);
		} else
			retval = mount(source, target, "", MS_BIND, NULL);
	}
	if (retval < 0)
		errExit("mount");
}

static int there_is_dash_dash(char *argv[]) {
	for (; *argv != NULL; argv++) {
		if (strcmp(argv[0], "--") == 0)
			return 1;
	}
	return 0;
}

int main(int argc, char *argv[]) {
	static char *short_options = "+shvn";
	static struct option long_options[] = {
		{"help", no_argument, 0, 'h'},
		{"newns", no_argument, 0, 'n'},
		{"sysadm", no_argument, 0, 's'},
		{"verbose", no_argument, 0, 'v'},
		{0, 0, 0, 0}};
	progname = basename(argv[0]);
	while (1) {
		int c;
		if ((c = getopt_long(argc, argv, short_options, long_options, NULL)) == -1)
			break;
		switch (c) {
			case 's': sysadmin_flag = 1; break;
			case 'v': verbose_flag = 1; break;
			case 'n': newnamespace_flag = 1; break;
			case '?':
			case 'h': 
			default: usage_n_exit();
		}
	}
	if (strcmp(argv[optind - 1], "--") == 0)
		optind--;
	argc -= optind;
	argv += optind;
	if (there_is_dash_dash(argv))
		newnamespace_flag = 1;
	if (newnamespace_flag) {
		if (verbose_flag)
			fprintf(stderr, "creating a user_namespace\n");
		if (userbindmount_unshare() < 0)
			errExit("unshare");
	}
	if (sysadmin_flag) {
		if (verbose_flag)  
			fprintf(stderr, "setting sysadm ambient capability\n");   
		if (userbindmount_set_cap_sysadm() < 0)
			errExit("set_cap_sysadm");
	}
	while (argc >= 2 && strcmp(argv[0],"--") != 0) {
		if (verbose_flag) 
			fprintf(stderr, "mounting %s on %s\n", argv[0], argv[1]);
		userbindmount_mount(argv[0], argv[1]);
		argc -= 2;
		argv += 2;
	}
	if (newnamespace_flag || argc > 0) {
		char **cmdargv;
		char *argvsh[]={getenv("SHELL"),NULL};
		if (argc > 0 && strcmp(argv[0],"--") == 0) {
			argc--;
			argv++;
		}
		cmdargv = (argc == 0) ? argvsh : argv;
		if (cmdargv[0] == NULL)
			errExit("$SHELL env variable not set\n");
		if (verbose_flag) 
			fprintf(stderr, "starting %s\n", cmdargv[0]);   
		execvp(cmdargv[0], cmdargv);
	}
}
